From 885f8ab517365295e1c0651702a85c1324eb35db Mon Sep 17 00:00:00 2001
From: Wim Taymans <wtaymans@redhat.com>
Date: Sun, 6 Oct 2024 12:36:18 +0200
Subject: [PATCH] jack: actually clear the mix io

When we are asked to clear the mix io areas, actually do it, otherwise
the process thread might still be accessing the old memory and crash.

Also check that we have set io on the port before we decrement the
counter with active io or else we have a negative value and cause
problems later. This can happen when we susupend and set io to NULL but
there was never any io set on the port.

Fixes #4337
---
 pipewire-jack/src/pipewire-jack.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/pipewire-jack/src/pipewire-jack.c b/pipewire-jack/src/pipewire-jack.c
index 1adfa2225..ccca29006 100644
--- a/pipewire-jack/src/pipewire-jack.c
+++ b/pipewire-jack/src/pipewire-jack.c
@@ -600,7 +600,9 @@ do_mix_set_io(struct spa_loop *loop, bool async, uint32_t seq,
 			port->global_mix->io[1] = &port->io[1];
 		}
 	} else {
-		if (--port->n_mix == 0 && port->global_mix != NULL) {
+		info->mix->io[0] = NULL;
+		info->mix->io[1] = NULL;
+		if (port->n_mix > 0 && --port->n_mix == 0 && port->global_mix != NULL) {
 			port->global_mix->io[0] = NULL;
 			port->global_mix->io[1] = NULL;
 		}
-- 
2.46.2

